File: output/xCodeEval_8000/C++/Java/6934ed89801e9ff3e9766db6b55d9968.java
import java.util.*;

class MCMF_CS2 {
    private int n;
    private int m;
    private long[] cap;
    private NODE[] nodes;
    private NODE sentinel_node;
    private ARC[] arcs;
    private ARC sentinel_arc;
    private BUCKET[] buckets;
    private BUCKET[] l_bucket;
    private long linf;
    private int time_for_price_in;
    private long epsilon;
    private long dn;
    private long price_min;
    private long mmc;
    private double f_scale;
    private double cut_off_factor;
    private double cut_on;
    private double cut_off;
    private long total_excess;
    private int flag_price;
    private int flag_updt;
    private int snc_max;
    private ARC d_arc;
    private NODE d_node;
    private NODE dummy_node;
    private long n_rel;
    private long n_ref;
    private long n_src;
    private long n_push;
    private long n_relabel;
    private long n_discharge;
    private long n_refine;
    private long n_update;
    private long n_scan;
    private long n_prscan;
    private long n_prscan1;
    private long n_prscan2;
    private long n_bad_pricein;
    private long n_bad_relabel;
    private long n_prefine;
    private boolean no_zero_cycles;
    private boolean check_solution;
    private boolean comp_duals;
    private boolean cost_restart;
    private boolean print_ans;
    private long[] node_balance;
    private long node_min;
    private long node_max;
    private long[] arc_first;
    private long[] arc_tail;
    private long pos_current;
    private ARC arc_current;
    private ARC arc_new;
    private ARC arc_tmp;
    private long max_cost;
    private long total_p;
    private long total_n;
    private NODE i_node;
    private NODE j_node;

    public MCMF_CS2(int num_nodes, int num_arcs) {
        n = num_nodes;
        m = num_arcs;

        flag_price = 0;
        flag_updt = 0;
        n_push = 0;
        n_relabel = 0;
        n_discharge = 0;
        n_refine = 0;
        n_update = 0;
        n_scan = 0;
        n_prscan = 0;
        n_prscan1 = 0;
        n_prscan2 = 0;
        n_bad_pricein = 0;
        n_bad_relabel = 0;
        n_prefine = 0;
        no_zero_cycles = false;
        check_solution = false;
        comp_duals = false;
        cost_restart = false;
        print_ans = true;

        allocateArrays();
    }

    private void errEnd(int cc) {
        System.exit(cc);
    }

    private void allocateArrays() {
        nodes = new NODE[n + 2];
        arcs = new ARC[2 * m + 1];
        cap = new long[2 * m];
        arc_tail = new long[2 * m];
        arc_first = new long[n + 2];

        for (int i = 0; i < n + 2; i++) {
            nodes[i] = new NODE();
        }

        if (nodes == null || arcs == null || arc_first == null || arc_tail == null) {
            errEnd(1);
        }

        pos_current = 0;
        arc_current = arcs;

        node_max = 0;
        node_min = n;
        max_cost = 0;
        total_p = 0;
        total_n = 0;
    }

    private void setArc(long tail_node_id, long head_node_id, long low_bound, long up_bound, long cost) {
        if (tail_node_id < 0 || tail_node_id > n || head_node_id < 0 || head_node_id > n) {
            errEnd(1);
        }
        if (up_bound < 0) {
            up_bound = Integer.MAX_VALUE;
        }
        if (low_bound < 0 || low_bound > up_bound) {
            errEnd(1);
        }

        arc_first[(int) tail_node_id + 1]++;
        arc_first[(int) head_node_id + 1]++;
        i_node = nodes[(int) tail_node_id];
        j_node = nodes[(int) head_node_id];

        arc_tail[(int) pos_current] = tail_node_id;
        arc_tail[(int) pos_current + 1] = head_node_id;
        arc_current.setHead(j_node);
        arc_current.setRezCapacity(up_bound - low_bound);
        cap[(int) pos_current] = up_bound;
        arc_current.setCost(cost);
        arc_current.setSister(arc_current + 1);
        (arc_current + 1).setHead(nodes[(int) tail_node_id]);
        (arc_current + 1).setRezCapacity(0);
        cap[(int) pos_current + 1] = 0;
        (arc_current + 1).setCost(-cost);
        (arc_current + 1).setSister(arc_current);

        i_node.decFirst();
        i_node.setFirst(i_node.first() - 1);
        j_node.decFirst();
        j_node.setFirst(j_node.first() - 1);

        i_node.setExcess(i_node.excess() - low_bound);
        j_node.setExcess(j_node.excess() + low_bound);

        if (head_node_id < node_min) {
            node_min = head_node_id;
        }
        if (tail_node_id < node_min) {
            node_min = tail_node_id;
        }
        if (head_node_id > node_max) {
            node_max = head_node_id;
        }
        if (tail_node_id > node_max) {
            node_max = tail_node_id;
        }

        if (cost < 0) {
            cost = -cost;
        }
        if (cost > max_cost && up_bound > 0) {
            max_cost = cost;
        }

        arc_current += 2;
        pos_current += 2;
    }

    private void setSupplyDemandOfNode(long id, long excess) {
        nodes[(int) id].setExcess(excess);
        if (excess > 0) {
            total_p += excess;
        }
        if (excess < 0) {
            total_n -= excess;
        }
    }

    private void preProcessing() {
        long i;
        long last, arc_num, arc_new_num;
        long tail_node_id;
        NODE head_p;
        ARC arc_new, arc_tmp;
        long up_bound;
        long cost;
        long cap_out;
        long cap_in;

        for (NODE in : nodes) {
            in.setFirst(arcs);
        }

        for (i = node_min + 1; i <= node_max + 1; i++) {
            arc_first[(int) i] += arc_first[(int) i - 1];
            nodes[(int) i].setFirst(arcs + arc_first[(int) i]);
        }

        for (i = node_min; i < node_max; i++) {
            last = nodes[(int) i + 1].first() - arcs;

            for (arc_num = arc_first[(int) i]; arc_num < last; arc_num++) {
                tail_node_id = arc_tail[(int) arc_num];

                while (tail_node_id != i) {
                    arc_new_num = arc_first[(int) tail_node_id];
                    arc_current = arcs + arc_num;
                    arc_new = arcs + arc_new_num;

                    head_p = arc_new.head();
                    arc_new.setHead(arc_current.head());
                    arc_current.setHead(head_p);

                    up_bound = cap[(int) arc_new_num];
                    cap[(int) arc_new_num] = cap[(int) arc_num];
                    cap[(int) arc_num] = up_bound;

                    up_bound = arc_new.rezCapacity();
                    arc_new.setRezCapacity(arc_current.rezCapacity());
                    arc_current.setRezCapacity(up_bound);

                    cost = arc_new.cost();
                    arc_new.setCost(arc_current.cost());
                    arc_current.setCost(cost);

                    if (arc_new != arc_current.sister()) {
                        arc_tmp = arc_new.sister();
                        arc_new.setSister(arc_current.sister());
                        arc_current.setSister(arc_tmp);

                        arc_current.sister().setSister(arc_current);

File: output/xCodeEval_remove_8000/C++/Go/6934ed89801e9ff3e9766db6b55d9968.go
package main

import (
	"fmt"
)

type id32 struct {
	n, m         int
	cap          []int
	nodes        []*node
	id31, id14   *node
	id30         *node
	arcs         []*arc
	id26         *arc
	buckets      []*bucket
	l_bucket     *bucket
	linf         int
	id16         int
	epsilon      int
	dn           int
	price_min    int
	mmc          int
	f_scale      float64
	id18         float64
	cut_on       float64
	cut_off      float64
	id23         int
	flag_price   int
	id29         int
	id0          int
	d_arc        arc
	d_node       node
	dummy_node   *node
	dnode        *node
	n_rel        int
	n_ref        int
	n_src        int
	n_push       int
	id24         int
	id17         int
	id22         int
	n_update     int
	n_scan       int
	id6          int
	id34         int
	id27         int
	id15         int
	id7          int
	id3          int
	id2          bool
	check_solution bool
	id1          bool
	cost_restart bool
	print_ans    bool
	node_balance []int
	node_min     int
	node_max     int
	arc_first    []int
	arc_tail     []int
	pos_current  int
	arc_current  *arc
	arc_new      *arc
	arc_tmp      *arc
	max_cost     int
	total_p      int
	total_n      int
	i_node       *node
	j_node       *node
}

type node struct {
	excess   int
	price    int
	first    *arc
	current  *arc
	suspended *arc
	q_next   *node
	b_next   *node
	b_prev   *node
	rank     int
	inp      int
}

type arc struct {
	id5     int
	cost    int
	head    *node
	sister  *arc
}

type bucket struct {
	p_first *node
}

func (g *id32) err_end(cc int) {
	fmt.Println(cc)
}

func (g *id32) allocate_arrays() {
	g.nodes = make([]*node, g.n+2)
	g.arcs = make([]*arc, 2*g.m+1)
	g.cap = make([]int, 2*g.m)
	g.arc_tail = make([]int, 2*g.m)
	g.arc_first = make([]int, g.n+2)

	for i := 0; i <= g.n; i++ {
		g.nodes[i] = &node{}
		g.nodes[i].excess = 0
	}

	if g.nodes == nil || g.arcs == nil || g.arc_first == nil || g.arc_tail == nil {
		g.err_end(1)
	}

	g.pos_current = 0
	g.arc_current = g.arcs

	g.node_max = 0
	g.node_min = g.n
	g.max_cost = 0
	g.total_p = 0
	g.total_n = 0
}

func (g *id32) deallocate_arrays() {
	if g.arcs != nil {
		g.arcs = nil
	}
	if g.dnode != nil {
		g.dnode = nil
	}
	if g.cap != nil {
		g.cap = nil
	}
	if g.buckets != nil {
		g.buckets = nil
	}
	if g.check_solution == true {
		g.node_balance = nil
	}
	if g.nodes != nil {
		g.nodes = g.nodes[g.node_min:]
		g.nodes = nil
	}
}

func (g *id32) set_arc(id10, id28, low_bound, up_bound, cost int) {
	if id10 < 0 || id10 > g.n || id28 < 0 || id28 > g.n {
		g.err_end(1)
	}
	if up_bound < 0 {
		up_bound = id9
	}
	if low_bound < 0 || low_bound > up_bound {
		g.err_end(1)
	}

	g.arc_first[id10+1]++
	g.arc_first[id28+1]++
	g.i_node = g.nodes[id10]
	g.j_node = g.nodes[id28]

	g.arc_tail[g.pos_current] = id10
	g.arc_tail[g.pos_current+1] = id28
	g.arc_current.head = g.j_node
	g.arc_current.rez_capacity = up_bound - low_bound
	g.cap[g.pos_current] = up_bound
	g.arc_current.cost = cost
	g.arc_current.sister = g.arc_current + 1
	g.arc_current.sister.head = g.nodes[id10]
	g.arc_current.sister.rez_capacity = 0
	g.cap[g.pos_current+1] = 0
	g.arc_current.sister.cost = -cost
	g.arc_current.sister.sister = g.arc_current

	g.i_node.excess -= low_bound
	g.j_node.excess += low_bound

	if id28 < g.node_min {
		g.node_min = id28
	}
	if id10 < g.node_min {
		g.node_min = id10
	}
	if id28 > g.node_max {
		g.node_max = id28
	}
	if id10 > g.node_max {
		g.node_max = id10
	}

	if cost < 0 {
		cost = -cost
	}
	if cost > g.max_cost && up_bound > 0 {
		g.max_cost = cost
	}

	g.arc_current += 2
	g.pos_current += 2
}

func (g *id32) set_supply_demand_of_node(id, excess int) {
	if id < 0 || id > g.n {
		g.err_end(1)
	}
	g.nodes[id].excess = excess
	if excess > 0 {
		g.total_p += excess
	}
	if excess < 0 {
		g.total_n -= excess
	}
}

func (g *id32) pre_processing() {
	g.arc_first[0] = 0

	for i := 0; i < g.n; i++ {
		g.arc_first[i+1] += g.arc_first[i]
		g.nodes[i+1].first = g.arcs + g.arc_first[i]
	}

	for i := 0; i < g.node_max; i++ {
		last := g.nodes[i+1].first - g.arcs
		for arc_num := g.arc_first[i]; arc_num < last; arc_num++ {
			id10 := g.arc_tail[arc_num]

			for id10 != i {
				id20 := g.arc_first[id10]
				g.arc_current = g.arcs + arc_num
				arc_new := g.arcs + id20

				head_p := arc_new.head
				arc_new.head = g.arc_current.head
				g.arc_current.head = head_p

				up_bound := g.cap[id20]
				g.cap[id20] = g.cap[arc_num]
				g.cap[arc_num] = up_bound

				up_bound = arc_new.rez_capacity
				arc_new.rez_capacity = g.arc_current.rez_capacity
				g.arc_current.rez_capacity = up_bound

				cost := arc_new.cost
				arc_new.cost = g.arc_current.cost
				g.arc_current.cost = cost

				if arc_new != g.arc_current.sister {
					arc_tmp := arc_new.sister
					arc_new.sister = g.arc_current.sister
					g.arc_current.sister = arc_tmp

					g.arc_current.sister.sister = g.arc_current
					arc_new.sister.sister = arc_new
				}

				g.arc_tail[arc_num] = g.arc_tail[id20]
				g.arc_tail[id20] = id10

				g.arc_first[id10]++

				id10 = g.arc_tail[arc_num]
			}
		}
	}

	g.n = g.node_max - g.node_min + 1
	g.nodes = g.nodes[g.node_min:]

	g.deallocate_arrays()
}

func (g *id32) run_cs2() {
	g.pre_processing()
	g.cs2_initialize()
	g.cs2_cost_restart(nil)
}

func main() {
	var n, m int
	fmt.Scan(&n, &m)
	grid := make([][]int, n)
	for i := 0; i < n; i++ {
		grid[i] = make([]int, m)
		for j := 0; j < m; j++ {
			fmt.Scan(&grid[i][j])
		}
	}
	a, b, c := make([]int, 0), make([]int, 0), make([]int, 0)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if (i+j)%2 == 1 {
				if i > 0 {
					a = append(a, f(i, j))
					b = append(b, f(i-1, j))
					c = append(c, grid[i][j] != grid[i-1][j])
				}
				if j > 0 {
					a = append(a, f(i, j))
					b = append(b, f(i, j-1))
					c = append(c, grid[i][j] != grid[i][j-1])
				}
				if i+1 < n {
					a = append(a, f(i, j))
					b = append(b, f(i+1, j))
					c = append(c, grid[i][j] != grid[i+1][j])
				}
				if j+1 < m {
					a = append(a, f(i, j))
					b = append(b, f(i, j+1))
					c = append(c, grid[i][j] != grid[i][j+1])
				}
			}
		}
	}
	numNodes := n * m * 4
	id13 := len(a)

	id8 := id32{n: numNodes, m: id13}
	for i := 0; i < id13; i++ {
		id8.set_arc(a[i], b[i], 0, 1, c[i])
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if (i+j)%2 == 1 {
				id8.set_supply_demand_of_node(f(i, j), 1)
			} else {
				id8.set_supply_demand_of_node(f(i, j), -1)
			}
		}
	}
	id8.run_cs2()
}
